File: platform/android/fips-ble/src/main/kotlin/org/fips/ble/FipsBleCommandRunner.kt
package org.fips.ble

interface FipsBlePlatform : AutoCloseable {
    fun setEventSink(sink: HostBleEventSink)

    fun submit(command: HostBleCommand)

    override fun close() = Unit
}

class FipsBleCommandRunner(
    private val platform: FipsBlePlatform,
    eventSink: HostBleEventSink = HostBleEventSink {},
) : AutoCloseable {
    init {
        platform.setEventSink(eventSink)
    }

    fun submit(command: HostBleCommand) {
        platform.submit(command)
    }

    override fun close() {
        platform.close()
    }
}

File: platform/apple/Sources/FipsBle/FipsBleCommandRunner.swift
public protocol FipsBlePlatform: AnyObject {
    var eventSink: HostBleEventSink { get set }

    func submit(_ command: HostBleCommand)

    func close()
}

public final class FipsBleCommandRunner {
    private let platform: FipsBlePlatform

    public init(
        platform: FipsBlePlatform,
        eventSink: @escaping HostBleEventSink = { _ in }
    ) {
        self.platform = platform
        platform.eventSink = eventSink
    }

    public func submit(_ command: HostBleCommand) {
        platform.submit(command)
    }

    public func close() {
        platform.close()
    }
}
